src/server/db_slice.cc

```diff
@@ -1096,10 +1109,79 @@ auto DbSlice::DeleteExpiredStep(const Context& cntx, unsigned count) -> DeleteEx
   return result;
 }
 
-// TODO: Design a better background evicting heuristic.
+int32_t DbSlice::GetNextSegmentForEviction(int32_t segment_id, DbIndex db_ind) const {
+  // wraps around if we reached the end
+  return db_arr_[db_ind]->prime.NextSeg((size_t)segment_id) %
+         db_arr_[db_ind]->prime.GetSegmentCount();
+}
+
 void DbSlice::FreeMemWithEvictionStep(DbIndex db_ind, size_t increase_goal_bytes) {
-  if (!caching_mode_)
+  DCHECK(!owner_->IsReplica());
+  if ((!caching_mode_) || !expire_allowed_ || !GetFlag(FLAGS_enable_heartbeat_eviction))
     return;
+
+  auto max_eviction_per_hb = GetFlag(FLAGS_max_eviction_per_heartbeat);
+  auto max_segment_to_consider = GetFlag(FLAGS_max_segment_to_consider);
+
+  auto time_start = absl::GetCurrentTimeNanos();
+  auto& db_table = db_arr_[db_ind];
+  int32_t num_segments = db_table->prime.GetSegmentCount();
+  int32_t num_buckets = PrimeTable::Segment_t::kTotalBuckets;
+  int32_t num_slots = PrimeTable::Segment_t::kNumSlots;
+
+  size_t used_memory_after;
+  size_t evicted = 0;
+  string tmp;
+  int32_t starting_segment_id = rand() % num_segments;
+  size_t used_memory_before = owner_->UsedMemory();
+  for (int32_t slot_id = num_slots - 1; slot_id >= 0; --slot_id) {
+    for (int32_t bucket_id = num_buckets - 1; bucket_id >= 0; --bucket_id) {
+      // pick a random segment to start with in each eviction,
+      // as segment_id does not imply any recency, and random selection should be fair enough
+      int32_t segment_id = starting_segment_id;
```

Comment:
I think random segment id has a disadvantage here as you go through max_segments_to_consider (default 4) in each time this function is called. So lets say the selection is distribute evenly but because you run in groups of 4 you will have overlap in segments between differet runs. f.e in the first run you starting segment was 1 leading to evicting from 1,2,3,4 and in the second run your starting segment was 4 leading to evicting from 4,5,6,7 there for you will be runnning on 4 in the first and the second time this function runs. 
I personally think savinge the last segment id is better than random, but you can also solve this by selecting (rand % (number_segments / max_segment_to_consider)) *  max_segment_to_consider i.e one of the segments from 0, 4, 8, 12...

Reply:
i actually tested something similar to this and they seems to perform very similarly with the current implementation... in the situation where eviction is likely needed, I found there are 1000+ segments... and each eviction is relatively short running (100 - 700us), so even when there is a collision, it will not collide again in the next eviction...

Reply:
While I agree there can be a bias in some cases, I think we can polish the logic in the subsequent iterations after we get feedback on how it works.